ManagedHooksManager/ApiHook.cs

```diff
@@ -0,0 +1,198 @@
+                addrOfFuncAddress,
+                eatFuncEntrySize,
+                NativeConstants.PAGE_WRITECOPY,
+                out var oldProtect);
+
+            if (!protectResult)
+                throw Marshal.GetExceptionForHR(Marshal.GetHRForLastWin32Error());
+
+            Marshal.WriteInt32(addrOfFuncAddress, newFuncRva);
+            GCHandle.Alloc(newFunc);
+
+            protectResult = NativeApi.VirtualProtect(
+                addrOfFuncAddress,
+                eatFuncEntrySize,
+                oldProtect,
+                out _);
+
+            if (!protectResult)
+                throw Marshal.GetExceptionForHR(Marshal.GetHRForLastWin32Error());
+
+            OriginalFunction = Marshal.GetDelegateForFunctionPointer<TDelegate>(
+                new IntPtr(moduleBaseAddress + originalFuncRva));
+        }
+
+        private bool TryFindExportNameOrdinal(
+            NativeStructs.IMAGE_EXPORT_DIRECTORY imageExportDirectory,
+            long moduleBaseAddress,
+            out short nameOrdinal,
+            out int functionRva)
+        {
+            var nameRvas = new int[imageExportDirectory.NumberOfNames];
+            Marshal.Copy(new IntPtr(moduleBaseAddress + imageExportDirectory.AddressOfNames), nameRvas, 0, nameRvas.Length);
+
+            var nameOrdinals = new short[imageExportDirectory.NumberOfNames];
+            Marshal.Copy(new IntPtr(moduleBaseAddress + imageExportDirectory.AddressOfNameOrdinals), nameOrdinals, 0, nameOrdinals.Length);
+
+            var funcAddrs = new int[imageExportDirectory.NumberOfFunctions];
+            Marshal.Copy(new IntPtr(moduleBaseAddress + imageExportDirectory.AddressOfFunctions), funcAddrs, 0, funcAddrs.Length);
+
+            nameOrdinal = -1;
+            functionRva = -1;
+
+            for (int i = 0; i < imageExportDirectory.NumberOfNames; i++)
+            {
+                var funcName = Marshal.PtrToStringAnsi(new IntPtr(moduleBaseAddress + nameRvas[i]));
+
+                if (string.Equals(funcName, _exportName, StringComparison.OrdinalIgnoreCase))
+                {
+                    nameOrdinal = nameOrdinals[i];
+                    functionRva = funcAddrs[nameOrdinal];
+                    break;
+                }
+            }
+
+            return nameOrdinal != -1;
+        }
+
+        private static int CreateX64Trampoline(long moduleBaseAddress, IntPtr newFuncPointer)
+        {
+            var memoryInfoStructSize = Marshal.SizeOf<NativeStructs.MEMORY_BASIC_INFORMATION64>();
+            var memoryInfoStructPtr = Marshal.AllocHGlobal(memoryInfoStructSize);
+
+            var trampolineAddress = IntPtr.Zero;
+            var trampoline = GetTrampolineInstructions(newFuncPointer);
+            var trampolineSize = new IntPtr(trampoline.Length);
+
+            const long tryMemoryStep = 0x10000L;
+
+            var tryMemoryBlock = moduleBaseAddress;
+            do
+            {
+                tryMemoryBlock += tryMemoryStep;
```

Comment:
Currently the place for the trampoline is being looked for within +2 GB of address space. This code also needs some checks to avoid allocating memory in system address space range.